nucleus/payara-modules/healthcheck-stuck/src/main/java/fish/payara/nucleus/healthcheck/stuck/StuckThreadsHealthCheck.java

```diff
@@ -59,58 +69,118 @@
+        acceptStuckThreads((workStartedTime, timeWorkingInMillis, thresholdInMillis, info) ->
+            result.add(new HealthCheckResultEntry(HealthCheckResultStatus.WARNING, "Stuck Thread: " + info.toString())));
+        return result;
+    }
 
-        Long thresholdNanos = TimeUnit.NANOSECONDS.convert(options.getTimeStuck(), options.getUnitStuck());
+    @Override
+    @MonitoringData(ns = "health", intervalSeconds = 4)
+    public void collect(MonitoringDataCollector collector) {
+        if (options == null || !options.isEnabled()) {
+            return;
+        }
+        AtomicInteger count = new AtomicInteger(0);
+        AtomicLong maxDuration = new AtomicLong(0L);
+        acceptStuckThreads((workStartedTime, timeWorkingInMillis, thresholdInMillis, info) -> {
+            String thread = info.getThreadName();
+            if (thread == null || thread.isEmpty()) {
+                thread = String.valueOf(info.getThreadId());
+            }
+            collector.annotate("StuckThreadDuration", timeWorkingInMillis, true, //
+                    "Thread", thread, // OBS! must be the first attribute as it is the key.
+                    "Started", String.valueOf(workStartedTime), //
+                    "Threshold", String.valueOf(thresholdInMillis), //
+                    "Locked", Boolean.toString(info.getLockInfo() != null), //
+                    "Suspended", String.valueOf(info.isSuspended()), //
+                    "State", composeStateText(info));
+            count.incrementAndGet();
+            maxDuration.updateAndGet(value -> Math.max(value, timeWorkingInMillis));
+        });
+        collector.collect("StuckThreadDuration", maxDuration);
+        collector.collect("StuckThreadCount", count);
+    }
 
+    @Override
+    public void collect(MonitoringWatchCollector collector) {
+        if (options == null || !options.isEnabled()) {
+            return;
+        }
+        collector.watch("ns:health StuckThreadDuration", "Stuck Threads", "ms")
+            .red(getThresholdInMillis(), -30000L, false, null, null, false);
+    }
+
+    public String composeStateText(ThreadInfo info) {
+        if (info.getLockInfo() == null) {
+            return "Running";
+        }
+        Thread.State state = info.getThreadState();
+        String action = state == State.BLOCKED ? "Blocked on " //
+                : state == State.WAITING || state == State.TIMED_WAITING ? "Waiting on " : "Running ";
+        return action + info.getLockInfo().toString();
+    }
+
+    private void acceptStuckThreads(StuckThreadConsumer consumer) {
+        ThreadMXBean bean = ManagementFactory.getThreadMXBean();
+        long thresholdInMillis = getThresholdInMillis();
+        long now = System.currentTimeMillis();
         ConcurrentHashMap<Long, Long> threads = stuckThreadsStore.getThreads();
-        for (Long thread : threads.keySet()){
-            Long timeHeld = threads.get(thread);
```

Comment:
NB. As discussed in chat the `timeHeld` here was a semantic confusion. The map contained the timestamp when the thread started the work. I changed the algorithm accordingly and also changed all the computation to be based on milliseconds as nanosecond level only would make sense if we would run the check often and fast enough (every t with a t < 1ms)

Reply:
I think this possibly needs a bit more cleanup, as you can still configure it with a threshold of 5 nanoseconds.

Particularly in the monitoring console you get a funny situation where you get a Threshold listed as `0` in comparison to something like `5ms`. In this case, you could possibly change it to have it say the threshold is `<1ms`

Reply:
Agreed, I think this should possibly be handled by a separate PR.

Reply:
👍 I'll create a separate Jira